src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs

```diff
@@ -63,6 +76,53 @@ public class QueryService : IQueryService
             return new QueryResourceResponse(responseMetadata);
         }
 
+        private Dictionary<QueryResource, HashSet<CustomTagFilterDetails>> GenerateQueryResourceToCustomTagMapping(IEnumerable<CustomTagStoreEntry> customTags)
+        {
+            Dictionary<QueryResource, HashSet<CustomTagFilterDetails>> ret = new Dictionary<QueryResource, HashSet<CustomTagFilterDetails>>();
+
+            foreach (CustomTagStoreEntry customTag in customTags)
+            {
+                DicomTag[] result;
+                DicomTag dicomTag;
+                if (customTag.Status.Equals(CustomTagStatus.Added) && _dicomTagPathParser.TryParse(customTag.Path, out result))
+                {
+                    dicomTag = result[0];
+
+                    if (customTag.Level.Equals(CustomTagLevel.Instance))
+                    {
+                        ret.TryAdd(QueryResource.AllInstances, new HashSet<CustomTagFilterDetails>());
+
+                        ret[QueryResource.AllInstances].Add(new CustomTagFilterDetails(customTag.Key, customTag.Level, dicomTag));
+
+                        ret.TryAdd(QueryResource.StudyInstances, new HashSet<CustomTagFilterDetails>());
+
+                        ret[QueryResource.StudyInstances].Add(new CustomTagFilterDetails(customTag.Key, customTag.Level, dicomTag));
+
+                        ret.TryAdd(QueryResource.StudySeriesInstances, new HashSet<CustomTagFilterDetails>());
+
+                        ret[QueryResource.StudySeriesInstances].Add(new CustomTagFilterDetails(customTag.Key, customTag.Level, dicomTag));
+                    }
+
+                    if (customTag.Level.Equals(CustomTagLevel.Instance) || customTag.Level.Equals(CustomTagLevel.Series))
+                    {
+                        ret.TryAdd(QueryResource.AllSeries, new HashSet<CustomTagFilterDetails>());
+
+                        ret[QueryResource.AllSeries].Add(new CustomTagFilterDetails(customTag.Key, customTag.Level, dicomTag));
```

Comment:
AllSeries should not contains customtag on instance level. 
https://github.com/microsoft/dicom-server/blob/fca9af145d1b183e18543d5e795635cb1036a38c/src/Microsoft.Health.Dicom.Core/Features/Query/QueryLimit.cs#L42
As defined above, AllSeriesTags is composed of AllStudiesTags and StudySeriesTags.

Reply:
Also need unit tests to make sure we are testing these calculations